File: python/30.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []
        
        word_len = len(words[0])
        window_len = len(words) * word_len
        word_count = {}
        for word in words:
            word_count[word] = word_count.get(word, 0) + 1
        
        res = []
        for i in range(len(s) - window_len + 1):
            cur_count = {}
            j = 0
            while j < window_len:
                cur_word = s[i+j:i+j+word_len]
                if cur_word not in word_count:
                    break
                cur_count[cur_word] = cur_count.get(cur_word, 0) + 1
                if cur_count[cur_word] > word_count[cur_word]:
                    break
                j += word_len
            if j == window_len:
                res.append(i)
        
        return res

File: C#/30.cs
public class Solution {
    public IList<int> FindSubstring(string s, string[] words) {
        IList<int> result = new List<int>();
        int wordLength = words[0].Length;
        int concatLength = wordLength * words.Length;
        Dictionary<string, int> wordCount = new Dictionary<string, int>();
        foreach(string word in words) {
            if(!wordCount.ContainsKey(word)) {
                wordCount[word] = 0;
            }
            wordCount[word]++;
        }
        for(int i = 0; i <= s.Length - concatLength; i++) {
            Dictionary<string, int> seen = new Dictionary<string, int>();
            int j = 0;
            while(j < words.Length) {
                string word = s.Substring(i + j * wordLength, wordLength);
                if(!wordCount.ContainsKey(word)) {
                    break;
                }
                if(!seen.ContainsKey(word)) {
                    seen[word] = 0;
                }
                seen[word]++;
                if(seen[word] > wordCount[word]) {
                    break;
                }
                j++;
            }
            if(j == words.Length) {
                result.Add(i);
            }
        }
        return result;
    }
}